### PR TITLE
OPS-970 HHS Provider, MultiAuth, FakeAuth, User Edit

### PR DESCRIPTION
**Pull request by tdonaworth**
## What changed

* Wired up HHS Authentication Provider
* Created `MultiAuth` Login schema
* Created `FakeAuth` Login for offline / dev / demo login
* Created User Edit page, mostly for `new user` profile confirmation
* Updated the `docker-compose` so that `backend` waits for `data-import` to comlete
* E2E no longer requires a different `docker-compose.e2e.yml` stack to be running
  * This will still require local DB cleanup, or re-running of `data-import` manually if run consecutively, since not all E2E tests do a full cleanup 😞 

## Issue

#970 
#963 
#1206 

### Test FakeAuth
1. Visit the homepage, you should be redirected to /login
2. Click "Sign in with FakeAuth"
3. Click "Admin User"
4. You should be logged in as "admin.demo@email.com"

### Test AMS Login (Requires PIV)
1. Click "Sign in with HHS AMS"
2. Should be redirected to AMS login
3. Login with PIV
4. Should be redirected back to the homepage, logged in as your user.

## Screenshots

image

image

image

## Links

TBD

### CODE REVIEW

File: backend/ops_api/ops/resources/users.py [hunk 1]
```diff
@@ -34,10 +47,57 @@ def get(self, id: int) -> Response:
         #    response = make_response({}, 401)  # nosemgrep
         #    return response
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/resources/users.py:49]: Do we still need commented code?

File: backend/ops_api/ops/resources/users.py [hunk 2]
```diff
@@ -34,10 +47,57 @@ def get(self, id: int) -> Response:
...
+            with OpsEventHandler(OpsEventType.UPDATE_USER) as meta:
+                old_user: User = User.query.get(id)
+                if not old_user:
+                    raise RuntimeError("Invalid User ID")
+                # schema = RequestBody.get_schema()
+                # OPSMethodView._validate_request(
+                #     schema=schema,
+                #     message=f"{message_prefix} - Params failed validation:",
+                # )
+
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/resources/users.py:64]: And here

File: backend/ops_api/ops/resources/users_schemas.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Optional
+from uuid import UUID
+
+# from marshmallow import validates_schema
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/resources/users_schemas.py:8]: Commented code

File: backend/ops_api/ops/utils/auth.py [hunk 4]
```diff
@@ -70,20 +74,57 @@ def create_oauth_jwt(
...
 
 
+def get_jwks(provider_metadata_url: str):
+    provider_uris = json.loads(
+        requests.get(
+            provider_metadata_url,
+            headers={"Accept": "application/json"},
+        ).content.decode("utf-8")
+    )
+    # current_app.logger.debug(f"********  provider_uris={provider_uris}")
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/utils/auth.py:100]: Commented out code.

File: backend/ops_api/ops/utils/auth.py [hunk 5]
```diff
@@ -70,20 +74,57 @@ def create_oauth_jwt(
...
+    provider: Optional[str] = None,
+) -> dict[str, str]:
+    # claims_options = {
+    #     "iss": {
+    #         "essential": True,
+    #         "values": current_app.config["AUTHLIB_OAUTH_CLIENTS"][provider]["client_id"],
+    #     },
+    #     "jti": {"validate": JWTClaims.validate_jti},
+    #     "exp": {"validate": JWTClaims.validate_exp},
+    # }
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/utils/auth.py:119]: More commented code.

File: backend/ops_api/ops/utils/authentication.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            "Authorization": f"Bearer {token}",
+            "Accept": "application/json",
+        }
+        try:
+            # current_app.logger.debug(f"get_user_info: header={header}")
+            user_jwt = requests.get(
+                self.user_info_url,
+                headers=header,
+            ).content.decode("utf-8")
+            # current_app.logger.debug(f"get_user_info: user_jwt_response={user_jwt}")
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/utils/authentication.py:235]: commented code

File: backend/ops_api/ops/utils/user.py [hunk 7]
```diff
@@ -11,25 +12,44 @@ class UserInfoDict(TypedDict):
...
+        current_app.logger.debug("Creating new user")
+        try:
+            # Find the role with the matching permission
+            role = current_app.db_session.query(Role).options(load_only(Role.name)).filter_by(name="unassigned").one()
+            user = User(
+                email=userinfo["email"],
+                oidc_id=userinfo["sub"],
+                # hhs_id=(userinfo["hhsid"] if userinfo["hhsid"] is not None else None),
+                # first_name=userinfo["given_name"],
+                # last_name=userinfo["family_name"],
```
Comment (xlorepdarkhelm) [inline backend/ops_api/ops/utils/user.py:32]: commented code

File: backend/ops_api/tests/ops/auth/test_oidc_controller.py [hunk 8]
```diff
@@ -16,22 +16,25 @@ def test_auth_post_fails(client):
...
+    data = {"code": "abc123"}  # missing provider
 
     res = client.post("/api/v1/auth/login/", json=data)
+    print(f"response: {res.json}")
     assert res.status_code == 400
 
-    event = mock_session.add.call_args[0][0]
-    assert event.event_status == OpsEventStatus.FAILED
+    # event = mock_session.add.call_args[0][0]
+    # assert event.event_status == OpsEventStatus.FAILED
```
Comment (xlorepdarkhelm) [inline backend/ops_api/tests/ops/auth/test_oidc_controller.py:37]: commented code

File: backend/ops_api/tests/ops/users/test_users.py [hunk 9]
```diff
@@ -64,3 +64,58 @@ def test_user_to_dict():
...
+
+# @pytest.mark.usefixtures("app_ctx")
+# def test_put_user_missing_data(auth_client):
+#     # Send a PUT request with missing data
+#     response = auth_client.put(
+#         f"/api/v1/users/4",
+#         json={"invalid_field": "Invalid Value"},
+#     )
+#     # Check that the response status code is 400 Bad Request
+#     assert response.status_code == 400
```
Comment (xlorepdarkhelm) [inline backend/ops_api/tests/ops/users/test_users.py:95]: commented code

File: frontend/src/components/Auth/AuthSection.jsx [hunk 10]
```diff
@@ -7,35 +7,38 @@ import { getAuthorizationCode } from "./auth";
...
+//     // TODO: Vefiry the Token!
+//     //const isValidToken = validateTooken(token);
+//     const decodedJwt = jwt_decode(token);
+//     const userId = decodedJwt["sub"];
+//     const userDetails = await getUserByOidc(userId);
 
-    dispatch(setUserDetails(userDetails));
-}
+//     dispatch(setUserDetails(userDetails));
+// }
```
Comment (xlorepdarkhelm) [inline frontend/src/components/Auth/AuthSection.jsx:22]: commented code

File: frontend/src/components/Auth/MultiAuthSection.jsx [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+// async function setActiveUser(token, dispatch) {
+//     // TODO: Vefiry the Token!
+//     //const isValidToken = validateToken(token);
+//     const decodedJwt = jwt_decode(token);
+//     const userId = decodedJwt["sub"];
+//     // eslint-disable-next-line react-hooks/rules-of-hooks
+//     const { data: user } = useGetUserByOIDCIdQuery(userId);
+
+//     dispatch(setUserDetails(user));
+// }
```
Comment (xlorepdarkhelm) [inline frontend/src/components/Auth/MultiAuthSection.jsx:21]: commented code

File: frontend/src/components/Auth/MultiAuthSection.jsx [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    const handleFakeAuthLogin = (user_type) => {
+        console.debug(`Logging in with FakeAuth: ${user_type}`);
+        sessionStorage.setItem("activeProvider", "fakeauth");
+        callBackend(user_type).catch(console.error);
+        //console.log(`API Login Response = ${JSON.stringify(fakeUsers[user].access_token)}`);
+        //localStorage.setItem("access_token", fakeUsers[user_type].access_token);
+        //console.log(`localStorage.getItem("access_token") = ${localStorage.getItem("access_token")}`);
+        //dispatch(login());
+        //setActiveUser(fakeUsers[user_type].access_token, dispatch);
+        //console.debug("FakeUser Logged In!");
```
Comment (xlorepdarkhelm) [inline frontend/src/components/Auth/MultiAuthSection.jsx:106]: commented code

File: frontend/src/components/Auth/auth.js [hunk 13]
```diff
@@ -23,7 +27,7 @@ export const logoutUser = async (stateToken) => {
     //   client_id=${CLIENT_ID}&
     //   post_logout_redirect_uri=${REDIRECT_URI}&
     //   state=abcdefghijklmnopabcdefghijklmnop
```
Comment (xlorepdarkhelm) [inline frontend/src/components/Auth/auth.js:29]: commented code

File: frontend/src/components/UI/Form/Input/Input.jsx [hunk 14]
```diff
@@ -28,17 +38,18 @@ const Input = ({ name, label = name, onChange, pending = false, messages = [], v
...
             />
         </div>
     );
 
-    function handleChange(e) {
-        onChange(name, e.target.value);
-    }
+    // function handleChange(e) {
+    //     onChange(name, e.target.value);
+    // }
```
Comment (xlorepdarkhelm) [inline frontend/src/components/UI/Form/Input/Input.jsx:52]: commented code

File: frontend/src/components/Users/UserInfoForm/EditUserForm.jsx [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+            // updateBudgetLineItemStatus({ id: bli.id, status: "UNDER_REVIEW" }).unwrap();
+            // const updatedUser = await updateUser({
+            //     id: formData.id,
+            //     first_name: formData.first_name,
+            //     last_name: formData.last_name,
+            //     email: formData.email,
+            //     division: formData.division,
+            //     oidc_id: formData.oidc_id,
+            //     hhs_id: formData.hhs_id,
+            // }).unwrap();
```
Comment (xlorepdarkhelm) [inline frontend/src/components/Users/UserInfoForm/EditUserForm.jsx:42]: commented code

File: frontend/src/components/Users/UserInfoForm/EditUserForm.jsx [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+            const updatedUser = await callBackend(`/api/v1/users/${formData.id}`, "PUT", formData);
+            // console.log(`Update Response Data: ${JSON.stringify(response)}`);
```
Comment (xlorepdarkhelm) [inline frontend/src/components/Users/UserInfoForm/EditUserForm.jsx:44]: here too

**Review by xlorepdarkhelm — CHANGES_REQUESTED** [review]
Looks pretty good overall, several pieces of commented code, one other note about PUT and PATCH for Users.

File: backend/ops_api/ops/resources/users.py [hunk 17]
```diff
@@ -1,14 +1,27 @@
-from flask import Response, request
+from typing import Any
+
+import marshmallow_dataclass as mmdc
+
+# from desert import schema
```
[the remark this answers is not in the document]
Comment (tdonaworth) [reply backend/ops_api/ops/resources/users.py:5]: intend on using schema validation, working through debug issues

**Comment by fpigeonjr** [comment]
🥩 this is a beefy PR
image

File: frontend/src/components/UI/NotificationCenter/NotificationCenter.jsx [hunk 18]
```diff
@@ -22,7 +22,7 @@ const NotificationCenter = () => {
         data: notifications,
         error,
         isLoading,
-    } = useGetNotificationsByUserIdQuery(userId, { pollingInterval: 5000 });
+    } = useGetNotificationsByUserIdQuery(userId, { pollingInterval: 500000 });
```
Comment (johndeange) [inline frontend/src/components/UI/NotificationCenter/NotificationCenter.jsx:25]: Do we want to revert this?

**Review by stevtek — APPROVED** [review]

File: .github/workflows/unit_test_reusable.yml [hunk 19]
```diff
@@ -43,15 +41,37 @@ jobs:
...
+      - name: Launch Stack
+        uses: ./.github/actions/run-full-stack
+        env:
+          JWT_PRIVATE_KEY: ${{ secrets.JWT_PRIVATE_KEY }}
+
       # Run the Cypress E2E Tests
       - name: E2E Test
         working-directory: frontend
-        run: yarn test:e2e
+        run: yarn test:e2e --spec ./cypress/e2e/loginPage.cy.js
```
[the remark this answers is not in the document]
Comment (tdonaworth) [reply .github/workflows/unit_test_reusable.yml:62]: good catch 👍

File: frontend/src/components/UI/NotificationCenter/NotificationCenter.jsx (hunk 18 above)
Comment (tdonaworth) [reply frontend/src/components/UI/NotificationCenter/NotificationCenter.jsx:25]: 👍

**Comment by tdonaworth** [comment]
Can't be sure, but still getting an `SSL_ERROR_HANDSHAKE_FAILURE_ALERT` on the HHS AMS Login. I saw this prior, and hoped it was temporary. Will create a ticket to see if this is something on the HHS side, or simply because we're trying to connect from an `http` localhost site.

image

**Comment by tdonaworth** [comment]
> Can't be sure, but still getting an `SSL_ERROR_HANDSHAKE_FAILURE_ALERT` on the HHS AMS Login. I saw this prior, and hoped it was temporary. Will create a ticket to see if this is something on the HHS side, or simply because we're trying to connect from an `http` localhost site.
> 
> image

Escalated to AMS Support: The AMS Helpdesk ticket number is 853834

**Review by johndeange — APPROVED** [review]

**Pull request merged**